panel_client/panel_client.py

```diff
@@ -0,0 +1,170 @@
+import struct
+import sys
+import time
+
+# Polling interval of the IO process
+MIN_LATENCY_MS = 10
+
+# characer that delimits socket messages
+DELIMITER_CHAR = '\r'
+
+
+class PanelStateBase:
+
+  def diff_states(old_state, new_state):
+    return {
+        k:v for k in new_state.iteritems()
+        if (k not in old_state) or (v != old_state[k])
+    }
+
+  def get_state_updates(self):
+    """Must implement this function for your panel.
+
+    Should return an iterable of control_id, state pairs for anything that has
+    changed since the last call.
+    """
+    raise NotImplementedError()
+
+  def get_controls(self):
+    """Must implement this function for your panel.
+
+    Should return the available controls.
+    """
+    raise NotImplementedError()
+
+  def display_message(self, message):
+    """Must implement this function for your panel.
+
+    Display a message from the server for the user."""
+
+
+def _make_update_message(update):
+  return {'message': 'set-state', 'data': {'id': update[0], 'state': update[1]}}
+
+
+def _make_announce_message(controls):
+  return {'message': 'announce', 'data': {'controls':controls}}
+
+
+def _panel_io_subprocess_main(panel_state_factory, action_queue, message_queue):
+  panel_state = panel_state_factory()
+  action_queue.put(_make_announce_message(panel_state.get_controls()))
+
+  while True:
+    for update in panel_state.get_state_updates():
+      if update is None:
+        # received shutdown signal
+        print("Bye-bye!")
+        return
+      action_queue.put(_make_update_message(update))
+    try:
+      panel_state.display_message(message_queue.get(block=False))
+    except EmptyQueueException:
+      pass
+    time.sleep(MIN_LATENCY_MS / 1000)
+
+
+def _server_io_subprocess_main(
+    action_queue,
+    message_queue,
+    messenger_factory=lambda: SpaceTeamMessenger(socket.socket)):
+  messenger = messenger_factory()
+  while True:
+    try:
+      action = action_queue.get(block=False)
+      if action is None:
+        # received shutdown signal
+        return
+      messenger.send(action)
+    except EmptyQueueException:
+      pass
+    for message in messenger.get_messages():
+      if message['message'] == 'display':
+        message_queue.put(message['data']['display'])
+    #time.sleep(MIN_LATENCY_MS / 1000)
+
+
+class PanelClient:
+  """Handles communication with the server and polling state updates."""
+
+  def __init__(self, panel_state_factory):
+    self._action_queue = multiprocessing.Queue()
+    self._message_queue = multiprocessing.Queue()
+    self._panel_state_factory = panel_state_factory
+
+  def start(self):
+    """Start I/O subprocess and begin communicating with server."""
+    self._panel_io_subprocess = multiprocessing.Process(
+        target=_panel_io_subprocess_main,
+        args=(self._panel_state_factory, self._action_queue, self._message_queue))
+    self._panel_io_subprocess.start()
+
+    self._server_io_subprocess = multiprocessing.Process(
```

Comment:
why are you putting this in a separate process?

Reply:
Also curious about this. Having the two separate loops above is kinda nice factoring but seems like they could run in threads or even in the same process (if the loops were consolidated) since neither of them block at the moment.

Reply:
You're right that there is no need to have two subprocesses. Just subprocessing the panel would be fine. Multiprocessing is nice in that it gets around the GIL. This structure also makes it easier to add fault recovery code that attempts to restart the panel if it crashes.